### PR TITLE
Manual testing on RTN14 and RTN15

### PR DESCRIPTION
**Pull request by tiholic**
This PR aims to acknowledge RTN14 and RTN15 are complied with in this package, however, few points are skipped considering nascent stage of this package.

> TL;DR at bottom of this

### RTN14a ✅ 

**Actions done**
1. Create realtime using an Invalid API Key
2. attach `connectionStateChange` listeners and `channelStateChange` listeners
3. connect to realtime instance using `realtime.connect()`

```
2020-07-31 17:31:29.424647 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 17:31:30.115235 : ConnectionStateChange event: ConnectionEvent.failed
                             Reason: ErrorInfo message=Invalid key in request: asdf_d231:fawefsd-afwesd. (See https://help.ably.io/error/40005 for help.) code=40005 statusCode=400 href=https://help.ably.io/error/40005
```

### RTN14b, RTN14g  ⛔ 
Ignoring as these spec items for now as they revolve around `ProtocolMessage`

### RTN14c, RTN14d, RTN14e, RTN14f ✅ 

**Turning wifi off** - reconnect triggers every 15 seconds

loops between states: `connecting` and `disconnected` every 15s for 60s (i.e., until the default `connectionStateTtl`)
there-further, state changes between `connecting` and `suspended` continuously.

```
2020-07-31 16:44:48.912345 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 16:44:48.965277 : ConnectionStateChange event: ConnectionEvent.disconnected
                             Reason: ErrorInfo message=sandbox-realtime.ably.io code=80000 statusCode=503 href=https://help.ably.io/error/80000
2020-07-31 16:45:03.950481 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 16:45:04.014063 : ConnectionStateChange event: ConnectionEvent.disconnected
                             Reason: ErrorInfo message=sandbox-realtime.ably.io code=80000 statusCode=503 href=https://help.ably.io/error/80000
2020-07-31 16:45:18.988102 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 16:45:19.062352 : ConnectionStateChange event: ConnectionEvent.disconnected
                             Reason: ErrorInfo message=sandbox-realtime.ably.io code=80000 statusCode=503 href=https://help.ably.io/error/80000
2020-07-31 16:45:34.020740 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 16:45:34.109471 : ConnectionStateChange event: ConnectionEvent.disconnected
                             Reason: ErrorInfo message=sandbox-realtime.ably.io code=80000 statusCode=503 href=https://help.ably.io/error/80000
2020-07-31 16:45:49.056232 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 16:45:49.127278 : ConnectionStateChange event: ConnectionEvent.suspended
                             Reason: ErrorInfo message=sandbox-realtime.ably.io code=80000 statusCode=503 href=https://help.ably.io/error/80000
2020-07-31 16:46:49.101562 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 16:46:49.178212 : ConnectionStateChange event: ConnectionEvent.suspended
                             Reason: ErrorInfo message=sandbox-realtime.ably.io code=80000 statusCode=503 href=https://help.ably.io/error/80000
2020-07-31 16:47:49.139349 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 16:47:49.213962 : ConnectionStateChange event: ConnectionEvent.suspended
                             Reason: ErrorInfo message=sandbox-realtime.ably.io code=80000 statusCode=503 href=https://help.ably.io/error/80000
```

#### Temporary key ✅ 

Created a temporary key on web console and deleted that after connection is established.
Connection disconnected, and eventually status changed to failed after 1 retry. 

```
2020-07-31 18:25:05.070597 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 18:25:05.296538 : ConnectionStateChange event: ConnectionEvent.connected
                             Reason: null
2020-07-31 18:25:12.123308 : ConnectionStateChange event: ConnectionEvent.disconnected
                             Reason: ErrorInfo message=Key/token status changed (revoke). (See https://help.ably.io/error/40131 for help.) code=40131 statusCode=401 href=https://help.ably.io/error/40131
2020-07-31 18:25:12.126130 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 18:25:12.408257 : ConnectionStateChange event: ConnectionEvent.failed
                             Reason: ErrorInfo message=Key revoked. (See https://help.ably.io/error/40131 for help.) code=40131 statusCode=401 href=https://help.ably.io/error/40131
```

#### Temporary app ✅ 

Created a temporary app on web console and deleted that after connection is established.
Connection disconnected, and eventually status changed to failed after 1 retry.

```
2020-07-31 18:14:26.818151 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 18:14:27.057491 : ConnectionStateChange event: ConnectionEvent.connected
                             Reason: null
2020-07-31 18:14:51.896204 : ConnectionStateChange event: ConnectionEvent.disconnected
                             Reason: ErrorInfo message=Application lEu0Vg disabled. (See https://help.ably.io/error/40300 for help.) code=40300 statusCode=403 href=https://help.ably.io/error/40300
2020-07-31 18:14:52.612866 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 18:14:52.926823 : ConnectionStateChange event: ConnectionEvent.failed
                             Reason: ErrorInfo message=Application lEu0Vg disabled. (See https://help.ably.io/error/40300 for help.) code=40300 statusCode=403 href=https://help.ably.io/error/40300
```

#### Restricted API key ✅ 

Trying to attach to a channel with an API key with restricted access

```
2020-07-31 19:40:59.586416 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 19:41:00.463867 : ConnectionStateChange event: ConnectionEvent.connected
                             Reason: null
//Here, trying to attach to channel that is not allowed with this API key
                             Attaching to channel test-channel: Current state ChannelState.initialized
                             ChannelStateChange: ChannelState.attaching
                             Reason: null
                             ChannelStateChange: ChannelState.suspended
                             Reason: ErrorInfo message=Channel denied access based on given capability; channelId = test-channel. (See https://help.ably.io/error/40160 for help.) code=40160 statusCode=401 href=https://help.ably.io/error/40160
                             Unable to attach to channel: ErrorInfo message=Channel denied access based on given capability; channelId = test-channel. (See https://help.ably.io/error/40160 for help.) code=40160 statusCode=401 href=https://help.ably.io/error/40160
```

#### Un-restricting the above API Key, and manually re-connecting ❓ 

The above, with updated API key settings to remove restrictions and trying a fresh connection again by `realtime.connect()` on same realtime instance:

```
2020-07-31 19:45:58.041846 : ConnectionStateChange event: ConnectionEvent.connecting
                             Reason: null
2020-07-31 19:45:58.315391 : ConnectionStateChange event: ConnectionEvent.connected
                             Reason: ErrorInfo message=Connection unavailable code=80002 statusCode=503 href=https://help.ably.io/error/80002
                             Attaching to channel test-channel: Current state ChannelState.suspended
```

⚠️ ⚠️ ⚠️ ⚠️ ⚠️ 
`ConnectionState` changes to `connected` but with a reason "Connection Unavailable" ❓ 
This sounds and feels fishy 🎣 

If a fresh realtime instance is used to connect, it connects as expected.

#### Another observation ❗ 

If a channel is already connected, it can publish and subscribe to/from that channel even if API Key settings are updated with new restricted permissions. This feels a bit concerning to me, but could be technically costly.

### RTN15a ❓ 

Token has expired => not sure how to create this scenario

### RTN15g (RTN15g1, RTN15g2, RTN15g3) ✅ 

These items are closely related with logs shared under heading `RTN14c, RTN14d, RTN14e, RTN14f` above

### RTN15i, RTN15b (RTN15b1, RTN15b2), RTN15c (RTN15c1, RTN15c2, RTN15c3, RTN15c4), RTN15e ⛔ 

Ignoring as these spec items for now as they revolve around `ProtocolMessage`

### RTN15f ⛔ 

Ignoring as these spec items for now as they revolve around `ACK` and `NACK`

### RTN15d ⛔ 

Integration Tests - unsure of the priority on integration tests at the moment.

### Temporary connection issues ✅ 

**Cause:**
1. bad network

**What's happening:** 
1. Automatic connection retry every 15 seconds
2. this connection goes into `disconnected` state until `connectionStateTtl` is reached
3. After `connectionStateTtl` is reached, connection goes into `suspended` state while connection re-establishment happens in background.

### Permanent connection issues ✅ 

**Cause:**

1. permanent API Key expiration (deletion of API key from web console)
2. permanent App expiration (deletion of app from web console)
3. Restricted API key (adding restrictions on API key)

**What's happening?**

1) On permanent API key expiration and (2) On permanent App expiration

`ConnectionState` changes in the following order and halts: `connecting` -> `connected` -> `disconnected` -> `connecting` -> `failed`

3) On restricting API Key

Connection succeeds initially, and when attaching to a channel that is restricted with the API key, it fails with code `40160` and `ChannelState` changes to `suspended` while `ConnectionState` still stays `connected`

**Additional Scenario:** ❓ 

 Un-restricting the restricted API Key - Connection on already available Realtime instance would not work. It requires us to use a fresh Realtime instance to be able to operate normally again.

### CODE REVIEW

File: example/lib/main.dart [hunk 1]
```diff
@@ -174,6 +174,10 @@ class _MyAppState extends State<MyApp> {
       listenRealtimeConnection(realtime);
       ably.RealtimeChannel channel = realtime.channels.get("test-channel");
       listenRealtimeChannel(channel);
+      setState(() {
+        _realtime = realtime;
+        _realtimeCreationState = OpState.Succeeded;
+      });
     } catch (error) {
```
Comment (zoechi) [inline example/lib/main.dart:181]: It's not really part of this PR but `realtime` doesn't seem to be used outside `try { ...}`
then there shouldn't be a reason to declare it outside `try { ... }`

I'd change

```dart
    ably.Realtime realtime;
    try {
      realtime = ably.Realtime(options: clientOptions);
```


```dart
    try {
      var realtime = ably.Realtime(options: clientOptions);
```

File: example/lib/main.dart [hunk 2]
```diff
@@ -183,85 +187,20 @@ class _MyAppState extends State<MyApp> {
...
-      omegaSubscription,
-      preZetaSubscription,
-      postZetaSubscription
-    ];
-
-    setState(() {
-      _realtime = realtime;
-      _realtimeCreationState = OpState.Succeeded;
-    });
+    connectionStateChangeSubscriptions = [ alphaSubscription ];
```
Comment (zoechi) [inline example/lib/main.dart:195]: This looks brittle to me. 

I'd instead change the field
```dart
List<StreamSubscription<ably.ConnectionStateChange>> connectionStateChangeSubscriptions;
```
to 
```
var _connectionStateChangeSubscriptions = <StreamSubscription<ably.ConnectionStateChange>>[];
```
and this code to 
```dart
_connectionStateChangeSubscriptions.add(alphaSubscription);
```

File: example/lib/main.dart [hunk 3]
```diff
@@ -377,18 +315,25 @@ class _MyAppState extends State<MyApp> {
...
+        if (typeCounter % 3 == 0) {
+          await _realtime.channels.get('test-channel').publish(
+            name: 'Hello', data: data);
+        } else if (typeCounter % 3 == 1) {
+          await _realtime.channels.get('test-channel').publish(message: m);
+        } else if (typeCounter % 3 == 2) {
+          await _realtime.channels.get('test-channel').publish(
+            messages: [m, m]);
+        }
+        if (realtimePublishCounter != 0 &&
```
Comment (zoechi) [inline example/lib/main.dart:328]: Just a suggestion to not repeat `typeCounter % 3 ==` multiple times
```dart
                  switch (typeCounter % 3) {
                    case 0:
                      await _realtime.channels
                          .get('test-channel')
                          .publish(name: 'Hello', data: data);
                      break;
                    case 1:
                      await _realtime.channels
                          .get('test-channel')
                          .publish(message: m);
                      break;
                    case 2:
                      await _realtime.channels
                          .get('test-channel')
                          .publish(messages: [m, m]);
                      break;
                  }
```

File: example/lib/neseted-realtime-events.dart [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import 'dart:async';
```
Comment (zoechi) [inline example/lib/neseted-realtime-events.dart:1]: should the file name be 
`nested...` instead of
`neseted...`?

Comment (zoechi) [reply example/lib/neseted-realtime-events.dart:1]: This file does not seem to be used anywhere. Is this intentional?

Comment (tiholic) [reply example/lib/neseted-realtime-events.dart:1]: Thanks for the typo find 😅 
I've moved some code that was being used for testing inside example app to this file to make `example/main.dart` a bit cleaner.

Comment (zoechi) [reply example/lib/neseted-realtime-events.dart:1]: shouldn't there be a button to resolve the conversation?

**Comment by tiholic** [comment]
Some comments from earlier PR's


https://github.com/ably/ably-flutter/pull/20#discussion_r464451601
> `var alphaSubscription = realtime.connection.on().listen(...)` should do.
Explicit type annotations for local variables is discouraged in Dart when the type can be inferred especially when it's obvious where it comes from.
This leads to less verbose code especially with long type names and generic type parameters as here.

File: example/lib/main.dart (hunk 2 above)
Comment (tiholic) [reply example/lib/main.dart]: addressed 86c4e6b0175a59c09dfd368dcc6130365c9f342d

File: example/lib/main.dart (hunk 1 above)
Comment (tiholic) [reply example/lib/main.dart:176]: addressed 933e6fddb31504e18fd257121406ff8195c14b06

File: example/lib/main.dart (hunk 3 above)
Comment (tiholic) [reply example/lib/main.dart:331]: Thanks! addressed 91c3389641f516dfc377e2499dc760c871ae5798

File: example/lib/neseted-realtime-events.dart (hunk 4 above)
Comment (tiholic) [reply example/lib/neseted-realtime-events.dart]: file renamed 8dff0397450cb085fdf75dc5635c8d99290e824d
status comment added c61e063ef001e6d2aaa3d81b76666c903b169fe9

Comment (tiholic) [reply example/lib/neseted-realtime-events.dart]: I can see a "Resolve conversation" button.

**Review by QuintinWillison — APPROVED** [review]
Minor comment. Happy for you to resolve, or leave as is, as you see fit.

File: ManualTesting.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
+This document aims to list our all the inputs from how some features are tested manually
+and feasible tests can be converted to automated tests. 
+
+This document currently targets **RTN14** and **RTN15**
+ sections of the [spec](https://docs.ably.io/client-lib-development-guide/features/#idl), partially
```
[the remark this answers is not in the document]
Comment (tiholic) [reply ManualTesting.md]: Done: 5c9c04ef5ee02c0d21fc495d2e5155bf14ad7191

**Review by zoechi — APPROVED** [review]

**Review by zoechi — APPROVED** [review]

**Comment by tiholic** [comment]
@QuintinWillison we can merge this and discuss on the results later as they are already added to ManualTesting.md
